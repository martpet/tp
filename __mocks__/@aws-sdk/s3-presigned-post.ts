export const createPresignedPost = vi.fn().mockResolvedValue('dummyPresignedPost');
