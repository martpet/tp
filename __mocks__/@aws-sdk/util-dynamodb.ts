export const unmarshall = vi
  .fn()
  .mockName('unmarshall')
  .mockReturnValue('dummyUnmarshalledItems');
