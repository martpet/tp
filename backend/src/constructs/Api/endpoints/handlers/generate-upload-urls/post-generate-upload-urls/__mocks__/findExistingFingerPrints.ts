export const findExistingFingerPrints = vi
  .fn()
  .mockName('findExistingFingerPrints')
  .mockResolvedValue(['dummyFingerprint1']);
