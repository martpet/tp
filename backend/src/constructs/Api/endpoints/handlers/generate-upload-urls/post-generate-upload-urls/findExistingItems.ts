import {
  BatchGetCommand,
  DynamoDBClient,
  DynamoDBDocument,
  photosTableOptions,
  unmarshall,
} from 'lambda-layer';

const ddbClient = new DynamoDBClient({});
const ddbDocClient = DynamoDBDocument.from(ddbClient);

export const findExistingItems = async (hashes: string[]) => {
  const { tableName, partitionKey } = photosTableOptions;
  const result: string[] = [];

  async function batchGet(keys: Record<string, any>[]) {
    const batchGetCommand = new BatchGetCommand({
      RequestItems: { [tableName]: { Keys: keys } },
    });

    const { Responses, UnprocessedKeys } = await ddbDocClient.send(batchGetCommand);

    if (Responses) {
      Responses[tableName].forEach(({ hash }) => result.push(hash));
    }

    if (UnprocessedKeys) {
      const nextKeys = UnprocessedKeys[tableName]?.Keys?.map((key) => unmarshall(key));

      if (nextKeys) {
        await batchGet(nextKeys);
      }
    }
  }

  const keys = hashes.map((hash) => ({ [partitionKey.name]: hash }));
  const maxItems = 100;
  const promises = [];

  for (let i = 0; i < keys.length; i += maxItems) {
    const chunk = keys.slice(i, i + maxItems);
    promises.push(batchGet(chunk));
  }

  await Promise.all(promises);

  return result;
};
