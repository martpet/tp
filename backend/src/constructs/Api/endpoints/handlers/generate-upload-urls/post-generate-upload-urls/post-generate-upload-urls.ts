import {
  APIGatewayProxyHandlerV2,
  ApiRouteHeaders,
  createPresignedPost,
  errorResponse,
  getIdTokenPayload,
  HandlerEnv,
  maxPhotoUploadSize,
  PostGenerateUploadUrlsRequest,
  PostGenerateUploadUrlsResponse,
  S3Client,
  StatusCodes,
} from 'lambda-layer';

import { findExistingFingerPrints } from './findExistingFingerPrints';

const s3Client = new S3Client({});

export const handler: APIGatewayProxyHandlerV2<PostGenerateUploadUrlsResponse> = async (
  event
) => {
  const { authorization } = event.headers as ApiRouteHeaders<'/settings'>;
  const { photoBucket } = process.env as HandlerEnv<'/generate-upload-urls', 'POST'>;
  let requestItems;
  let existingFingerprintsInDb: string[];

  if (!authorization) {
    return errorResponse('Vf5Ph6qN1S');
  }

  if (!photoBucket) {
    return errorResponse('1rQkj3kpp4');
  }

  if (!event.body) {
    return errorResponse('G-luuHqI0s', { statusCode: StatusCodes.BAD_REQUEST });
  }

  try {
    requestItems = JSON.parse(event.body) as PostGenerateUploadUrlsRequest;
  } catch (error) {
    return errorResponse('9210145fdf', { statusCode: StatusCodes.BAD_REQUEST, error });
  }

  try {
    const fingerprints = requestItems.map(({ fingerprint }) => fingerprint);
    existingFingerprintsInDb = await findExistingFingerPrints(fingerprints);
  } catch (error) {
    return errorResponse('9fb96f4182', { error });
  }

  const uniqueItems = requestItems.filter(
    ({ fingerprint }) => !existingFingerprintsInDb.includes(fingerprint)
  );
  const { sub } = await getIdTokenPayload(authorization);

  const uploadUrlsEntries = await Promise.all(
    uniqueItems.map(async ({ id, fingerprint, digest }) => {
      const presignedPost = await createPresignedPost(s3Client, {
        Bucket: photoBucket,
        Key: `${sub}/${fingerprint}.jpg`,
        Expires: uniqueItems.length * 10,
        Fields: {
          'x-amz-checksum-algorithm': 'SHA256',
          'x-amz-checksum-sha256': digest,
        },
        Conditions: [['content-length-range', 0, maxPhotoUploadSize]],
      });
      return [id, presignedPost];
    })
  );

  return {
    presignedPosts: Object.fromEntries(uploadUrlsEntries),
    existingFingerprintsInDb,
  };
};
