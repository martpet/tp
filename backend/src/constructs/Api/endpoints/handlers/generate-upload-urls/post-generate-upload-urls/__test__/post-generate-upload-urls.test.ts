import { createPresignedPost } from '@aws-sdk/s3-presigned-post';

import {
  itCalls,
  itGetsIdToken,
  itHasEnvVars,
  itHasJsonBody,
  itResolves,
  itResolvesWithError,
} from '~/constructs/Api/utils';

import { findExistingFingerPrints } from '../findExistingFingerPrints';
import { handler } from '../post-generate-upload-urls';

vi.mock('@aws-sdk/s3-presigned-post');
vi.mock('@aws-sdk/client-s3');
vi.mock('~/constructs/Api/utils/errorResponse');
vi.mock('~/constructs/Api/utils/getIdTokenPayload');
vi.mock('../findExistingFingerPrints');

process.env.photoBucket = 'dummyPhotoBucket';

const args = [
  {
    headers: { authorization: 'dummyAuthorizationHeader' },
    body: JSON.stringify([
      {
        id: 'dummyId1',
        fingerprint: 'dummyFingerprint1',
        hash: 'dummyHash1',
      },
      {
        id: 'dummyId2',
        fingerprint: 'dummyFingerprint2',
        hash: 'dummyHash2',
      },
    ]),
  },
] as unknown as Parameters<typeof handler>;

describe('post-generate-upload-urls', () => {
  itHasJsonBody(handler, args);
  itHasEnvVars(['photoBucket'], handler, args);
  itGetsIdToken(handler, args);
  itResolves(handler, args);
  itCalls(createPresignedPost, handler, args);
  itCalls(findExistingFingerPrints, handler, args);

  describe('when `findExistingFingerPrints` throws', () => {
    beforeEach(() => {
      const error = new Error('dummyFindExistingItemsError');
      vi.mocked(findExistingFingerPrints).mockRejectedValueOnce(error);
    });
    itResolvesWithError(handler, args);
  });
});
