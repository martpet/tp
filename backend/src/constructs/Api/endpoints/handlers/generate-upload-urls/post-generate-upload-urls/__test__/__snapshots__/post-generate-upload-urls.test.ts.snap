// Vitest Snapshot v1

exports[`post-generate-upload-urls > calls "createPresignedPost" with correct args 1`] = `
[
  [
    DummyS3Client {},
    {
      "Bucket": "dummyPhotoBucket",
      "Conditions": [
        [
          "content-length-range",
          0,
          31457280,
        ],
      ],
      "Expires": 30,
      "Key": "dummySub/2022-01-01T00:00:00.000Z/dummyFingerprint2.jpg",
    },
  ],
]
`;

exports[`post-generate-upload-urls > calls "findExistingItems" with correct args 1`] = `
[
  [
    [
      "dummyFingerprint1",
      "dummyFingerprint2",
    ],
  ],
]
`;

exports[`post-generate-upload-urls > calls "getIdTokenPayload" with correct args 1`] = `
[
  [
    "dummyAuthorizationHeader",
  ],
]
`;

exports[`post-generate-upload-urls > resolves with a correct value 1`] = `
{
  "existingFingerprintsInDb": [
    "dummyFingerprint1",
  ],
  "uploadUrls": {
    "dummyFingerprint2": "dummyPresignedPost",
  },
}
`;

exports[`post-generate-upload-urls > when "authorization" header is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "Vf5Ph6qN1S",
  ],
]
`;

exports[`post-generate-upload-urls > when "authorization" header is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`post-generate-upload-urls > when "event.body" is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "G-luuHqI0s",
    {
      "statusCode": 400,
    },
  ],
]
`;

exports[`post-generate-upload-urls > when "event.body" is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`post-generate-upload-urls > when "event.body" is not JSON > calls "errorResponse" with correct args 1`] = `
[
  [
    "9210145fdf",
    {
      "error": [SyntaxError: Unexpected token o in JSON at position 1],
      "statusCode": 400,
    },
  ],
]
`;

exports[`post-generate-upload-urls > when "event.body" is not JSON > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`post-generate-upload-urls > when "photoBucket" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "1rQkj3kpp4",
  ],
]
`;

exports[`post-generate-upload-urls > when "photoBucket" env var is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`post-generate-upload-urls > when \`findExistingItems\` throws > calls "errorResponse" with correct args 1`] = `
[
  [
    "9fb96f4182",
    {
      "error": [Error: dummyFindExistingItemsError],
    },
  ],
]
`;

exports[`post-generate-upload-urls > when \`findExistingItems\` throws > resolves with a correct value 1`] = `"dummyErrorResponse"`;
